README.md
# The NIMH Healthy Research Volunteer Dataset

A comprehensive clinical, MRI, and MEG collection characterizing healthy research volunteers collected at the National Institute of Mental Health (NIMH) Intramural Research Program (IRP) in Bethesda, Maryland using medical and mental health assessments, diagnostic and dimensional measures of mental health, cognitive and neuropsychological functioning, structural and functional magnetic resonance imaging (MRI), along with diffusion tensor imaging (DTI), and a comprehensive magnetoencephalography battery (MEG).

In addition, blood samples are currently banked for future genetic analysis. All data collected in this protocol are broadly shared in the OpenNeuro repository, in the Brain Imaging Data Structure (BIDS) format. In addition, blood samples of healthy volunteers are banked for future analyses. All data collected in this protocol are broadly shared here, in the Brain Imaging Data Structure (BIDS) format. In addition, task paradigms and basic pre-processing scripts are shared on GitHub. This dataset is unique in its depth of characterization of a healthy population in terms of brain health and will contribute to a wide array of secondary investigations of non-clinical and clinical research questions.

This dataset is licensed under the [Creative Commons Zero (CC0) v1.0 License](https://creativecommons.org/publicdomain/zero/1.0/).

## Recruitment

Inclusion criteria for the study require that participants are adults at or over 18 years of age in good health with the ability to read, speak, understand, and provide consent in English. All participants provided electronic informed consent for online screening and written informed consent for all other procedures. Exclusion criteria include:

- A history of significant or unstable medical or mental health condition requiring treatment
- Current self-injury, suicidal thoughts or behavior
- Current illicit drug use by history or urine drug screen
- Abnormal physical exam or laboratory result at the time of in-person assessment
- Less than an 8th grade education or IQ below 70
- Current employees, or first-degree relatives of NIMH employees

Study participants are recruited through direct mailings, bulletin boards and listservs, outreach exhibits, print advertisements, and electronic media.

## Clinical Measures

All potential volunteers first visit the study website ([https://nimhresearchvolunteer.ctss.nih.gov](https://nimhresearchvolunteer.ctss.nih.gov)), check a box indicating consent, and complete preliminary self-report screening questionnaires. The study website is HIPAA compliant and therefore does not collect PII ; instead, participants are instructed to contact the study team to provide their identity and contact information. The questionnaires include demographics, clinical history including medications, disability status (WHODAS 2.0), mental health symptoms (modified DSM-5 Self-Rated Level 1 Cross-Cutting Symptom Measure), substance use survey (DSM-5 Level 2), alcohol use (AUDIT), handedness (Edinburgh Handedness Inventory), and perceived health ratings. At the conclusion of the questionnaires, participants are again prompted to send an email to the study team. Survey results, supplemented by NIH medical records review (if present), are reviewed by the study team, who determine if the participant is likely eligible for the protocol. These participants are then scheduled for an in-person assessment. Follow-up phone screenings were also used to determine if participants were eligible for in-person screening.

## In-person Assessments

At this visit, participants undergo a comprehensive clinical evaluation to determine final eligibility to be included as a healthy research volunteer. The mental health evaluation consists of a psychiatric diagnostic interview (Structured Clinical Interview for DSM-5 Disorders (SCID-5), along with self-report surveys of mood (Beck Depression Inventory-II (BD-II) and anxiety (Beck Anxiety Inventory, BAI) symptoms. An intelligence quotient (IQ) estimation is determined with the Kaufman Brief Intelligence Test, Second Edition (KBIT-2). The KBIT-2 is a brief (20-30 minute) assessment of intellectual functioning administered by a trained examiner. There are three subtests, including verbal knowledge, riddles, and matrices.

## Medical Evaluation

Medical evaluation includes medical history elicitation and systematic review of systems. Biological and physiological measures include vital signs (blood pressure, pulse), as well as weight, height, and BMI. Blood and urine samples are taken and a complete blood count, acute care panel, hepatic panel, thyroid stimulating hormone, viral markers (HCV, HBV, HIV), C-reactive protein, creatine kinase, urine drug screen and urine pregnancy tests are performed. In addition, blood samples that can be used for future genomic analysis, development of lymphoblastic cell lines or other biomarker measures are collected and banked with the NIMH Repository and Genomics Resource (Infinity BiologiX). The Family Interview for Genetic Studies (FIGS) was later added to the assessment in order to provide better pedigree information; the Adverse Childhood Events (ACEs) survey was also added to better characterize potential risk factors for psychopathology. The entirety of the in-person assessment not only collects information relevant for eligibility determination, but it also provides a comprehensive set of standardized clinical measures of volunteer health that can be used for secondary research.

## MRI Scan

Participants are given the option to consent for a magnetic resonance imaging (MRI) scan, which can serve as a baseline clinical scan to determine normative brain structure, and also as a research scan with the addition of functional sequences (resting state and diffusion tensor imaging). The MR protocol used was initially based on the ADNI-3 basic protocol, but was later modified to include portions of the ABCD protocol in the following manner:

1. The T1 scan from ADNI3 was replaced by the T1 scan from the ABCD protocol.
2. The Axial T2 2D FLAIR acquisition from ADNI2 was added, and fat saturation turned on.
3. Fat saturation was turned on for the pCASL acquisition.
4. The high-resolution in-plane hippocampal 2D T2 scan was removed and replaced with the whole brain 3D T2 scan from the ABCD protocol (which is resolution and bandwidth matched to the T1 scan).
5. The slice-select gradient reversal method was turned on for DTI acquisition, and reconstruction interpolation turned off.
6. Scans for distortion correction were added (reversed-blip scans for DTI and resting state scans).
7. The 3D FLAIR sequence was made optional and replaced by one where the prescription and other acquisition parameters provide resolution and geometric correspondence between the T1 and T2 scans.

At the time of the MRI scan, volunteers are administered a subset of tasks from the NIH Toolbox Cognition Battery. The four tasks include:

1. Flanker inhibitory control and attention task assesses the constructs of attention and executive functioning.
2. Executive functioning is also assessed using a dimensional change card sort test.
3. Episodic memory is evaluated using a picture sequence memory test.
4. Working memory is evaluated using a list sorting test.

## MEG

An optional MEG study was added to the protocol approximately one year after the study was initiated, thus there are relatively fewer MEG recordings in comparison to the MRI dataset. MEG studies are performed on a 275 channel CTF MEG system (CTF MEG, Coquiltam BC, Canada). The position of the head was localized at the beginning and end of each recording using three fiducial coils. These coils were placed 1.5 cm above the nasion, and at each ear, 1.5 cm from the tragus on a line between the tragus and the outer canthus of the eye. For 48 participants (as of 2/1/2022), photographs were taken of the three coils and used to mark the points on the T1 weighted structural MRI scan for co-registration. For the remainder of the participants (n=16 as of 2/1/2022), a Brainsight neuronavigation system (Rogue Research, Montréal, Québec, Canada) was used to coregister the MRI and fiducial localizer coils in realtime prior to MEG data acquisition.

## Specific Measures within Dataset

Online and In-person behavioral and clinical measures, along with the corresponding phenotype file name, sorted first by measurement location and then by file name.

| Location  | Measure                                                                                         | File Name                |
| --------- | ----------------------------------------------------------------------------------------------- | ------------------------ |
| Online    | Alcohol Use Disorders Identification Test (AUDIT)                                               | audit                    |
|           | Demographics                                                                                    | demographics             |
|           | DSM-5 Level 2 Substance Use - Adult                                                             | drug_use                 |
|           | Edinburgh Handedness Inventory (EHI)                                                            | ehi                      |
|           | Health History Form                                                                             | health_history_questions |
|           | Perceived Health Rating - self                                                                  | health_rating            |
|           | DSM-5 Self-Rated Level 1 Cross-Cutting Symptoms Measure – Adult (modified)                      | mental_health_questions  |
|           | World Health Organization Disability Assessment Schedule 2.0 (WHODAS 2.0)                       | whodas                   |
| In-Person | Adverse Childhood Experiences (ACEs)                                                            | ace                      |
|           | Chemistry Panel                                                                                 | acute_care               |
|           | Beck Anxiety Inventory (BAI)                                                                    | bai                      |
|           | Beck Depression Inventory-II (BDI-II)                                                           | bdi                      |
|           | Creatine Kinase, C-reactive Protein, TSH                                                        | blood_chemistry          |
|           | Complete Blood Count with Differential                                                          | cbc_with_differential    |
|           | Medical and mental health diagnosis, medications, physical exam, lab findings, vital signs, BMI | clinical_variable_form   |
|           | Family Interview for Genetic Studies (FIGS)                                                     | figs                     |
|           | Hepatic Panel                                                                                   | hepatic                  |
|           | Kaufman Brief Intelligence Test 2nd Edition (KBIT-2) and Visual Analogue of Effort Scale (VAS)  | kbit2_vas                |
|           | MRI Variables form                                                                              | mri_variables            |
|           | NIH Toolbox measures                                                                            | nih_toolbox              |
|           | Viral markers (Hepatitis B, C and HIV)                                                          | other                    |
|           | Perceived Health Rating - clinician                                                             | perceived_health_rating  |
|           | Research participation satisfaction survey                                                      | satisfaction             |
|           | Structured Clinical Interview for DSM-5 Disorders (SCID-5)                                      | scid5                    |
|           | Urine drug screen and pregnancy test (if indicated)                                             | urine_chemistry          |

## Data Preparation Notes

In many of the Clinical Measures data files, there exist some `-999` values. `-999` means there was no response though a response was possible. The question may have been skipped over by the participant or the question flow. `-777` appears in the Edinburgh Handedness Inventory (EHI) as well. `-777` means there is no data available for a response. The question was not presented or asked to the participant.

The data were prepared using the following tools and filename mappings.

### Biological and Physiological Measures Data

The `01_cris_clean_up.py` python script contains the functions used to clean and convert the spreadsheet with clinical measures to BIDS-standard TSV files and their data dictionaries to BIDS-standard JSON files.

### Clinical Measures Data

The `02_ctdb_clean_up.py` python script contains the functions used to clean and convert the spreadsheet downloaded from CTDB to BIDS-standard TSV files as well as their respective data dictionaries converted to BIDS-standard JSON files.

Some of the python functions common to both scripts were defined in `rvoldefinitions.py`

### BIDS-standard MEG Files

Data collected by the NIMH MEG Core was converted to BIDS-standard files using the MNE BIDS package. Associated notebooks: `1_mne_bids_extractor.ipynb` & `2_bids_editor.ipynb`. In addition, event codes and correct timings were generated with `hv_process.py` script, and the fiducial locations from the AFNI HEAD files were generated using the `export_tags.py` script. 

### BIDS-standard MRI

We used the `heudiconv` tool to convert MRI DICOM files to BIDS-standard files with the associated script: `hv_protocol_heuristic.py`. A modified workflow of `pydeface` was used to deface structural scans with the associated python script: `hv_protocol_pydeface_workflow.py`

Each participant received either the ADNI3 or the ABCD protocol, not both, during their MRI/MEG visit. T1w scans with acquisition label `fspgr` are ADNI3 protocol sequence and scans with `mprage` acquisition labels are ABCD protocol sequence.

### OpenNeuro BIDS File/Folder Tree

Below is a BIDS-compliant file/folder tree as it appears for subjects on OpenNeuro.

```shell
sub-ON<subject>
    └── ses-01
        ├── anat
        │   └── sub-ON<subject>_ses-01_acq-<desc>_run-<index>_<suffix>.<json|nii.gz>
        ├── asl
        │   └── sub-ON<subject>_ses-01_run-<index>_asl.<json|nii.gz>
        ├── dwi
        │   └── sub-ON<subject>_ses-01_run-<index>_dwi.<bvec|bval|json|nii.gz>
        ├── fmap
        │   └── sub-ON<subject>_ses-01_acq-<desc>_dir-<direction>_run-<index>_epi.<bvec|bval|json|nii.gz>
        ├── func
        │   └── sub-ON<subject>_ses-01_task-<taskname>_run-<index>_<suffix>.<json|nii.gz>
        ├── meg
        │   ├── sub-ON<subject>_ses-01_task-<taskname>_run-01_<meg|coordsystem>.json
        │   ├── sub-ON<subject>_ses-01_task-<taskname>_run-01_<channels|events>.tsv
        │   └── sub-ON<subject>_ses-01_task-<taskname>_run-01_meg.ds
        │       ├── BadChannels
        │       ├── bad.segments
        │       ├── ClassFile.cls
        │       ├── MarkerFile.mrk
        │       ├── params.dsc
        │       ├── processing.cfg
        │       ├── sub-ON<subject>_ses-01_task-<taskname>_run-01_meg.<extension>
        │       └── sub-ON<subject>_ses-01_task-<taskname>_run-01.xml
        └── sub-ON<subject>_ses-01_scans.<json|tsv>
```

Definitions:

- `<subject>` = subject number
- `<taskname>` = task name: `airpuff`, `artifact`, `gonogo`, `haririhammer`, `movie`, `oddball`, `sternberg`
- `<desc>` = placeholder for acquisition label for a given suffix
- `<direction>` = flipped, unflipped
- `<index>` = run number/index
- `<suffix>` = placeholder to indicate the scan type
  - `T1w`: `<desc>` = `fspgr`, `mprage`, `fse`, `highreshippo`
  - `T2w`: `<desc>` = `abcdcube`, `cube`, `frfse`
  - `FLAIR`: `<desc>` = `adni2d`, `2d`, `3d`, `t2`
  - `epi`: `<desc>` = `dwib1000`, `dwi`, `resting`
  - `T2star`
  - `bold`
  - `meg`
  - `asl`
- `<extension>`: indicates meg data files' type = `acq`, `bak`, `hc`, `hist`, `infods`, `meg4`, `newds`, `res4`, `xml`


CHANGES
1.0.0 2022-03-01
  - Initial release 


phenotype_data_prep_scripts/rvoldefinitions.py
import csv
import json
from collections import defaultdict
from os import fspath
from pathlib import Path

import pandas as pd

import project_filepaths as filepaths

# reading id_linking_file as a data frame
id_mapping = pd.read_csv(filepaths.id_filepath, low_memory=False)
rvs_to_be_shared = id_mapping['SUBJECT_NUMBER'].tolist()

# participants to be excluded from CTDB and CRIS data
combined_forms = ['ehi.tsv', 'ehi_r.tsv', 'whodas.tsv', 'whodas_r.tsv', 'demographics.tsv', 'demographics_r.tsv']


# definitions used for data cleaning
def openneuro_id_lookup(rvid):
    """given a rvid return the corresponding openneuro id"""
    onid = id_mapping.loc[id_mapping['SUBJECT_NUMBER'] == rvid, 'open_neuro_id'].values[0]
    return onid


def rvid_lookup(onid):
    """given an openneuro_id return the corresponding rvid"""
    rvid = id_mapping.loc[id_mapping['open_neuro_id'] == onid, 'SUBJECT_NUMBER'].values[0]
    return rvid


def duplicate_onids():
    # TODO: load openneuroids from id linking file and verify that there are no duplicates
    # returns either an empty list or non empty list with rvids that have the same
    # openneuro id
    # this function will be called after a new id linking file has been created
    # and before regenerating tsvs
    return None


def duplicate_rvids():
    # TODO: load rvid from id linking file and verify that there are no duplicates
    return None


def new_subjs_check():
    # TODO: find a list of all subjects in the new downloads
    # TODO: compare it with subjects available in participants.tsv
    # TODO: if any new subjects, then create a new id linking file with the new subjects
    return None


def remove_nans(dict_of_dfs):
    """replaces NaNs from a dictionary of dataframes"""
    for key, df in dict_of_dfs.items():
        dict_of_dfs[key] = dict_of_dfs[key].fillna(-999)
    return dict_of_dfs


def reorder_cols(df):
    """places the `participant_id` column as the first column"""
    # print(df.columns)
    req_order = [df.columns.tolist()[-1]] + df.columns.tolist()[:-1]
    # print(req_order)
    df = df.loc[:, req_order]
    df = df.sort_values(by=['participant_id'], ignore_index=True)
    return df


def remove_blank_rows(df):
    """remove participants that have no data available aka blank rows"""
    clean_df = df.filter(df.columns[1:], axis=1).dropna(how='all', axis=0)
    idx_to_keep = clean_df.index
    clean_df = df.filter(idx_to_keep, axis=0)
    return clean_df


def multiple_subj_responses(modality):
    if modality == 'cris':
        tsvs = list(filepaths.cris_out_dir.glob('*.tsv'))
    elif modality == 'ctdb':
        tsvs = list(filepaths.ctdb_out_dir.glob('*.tsv'))

    d = defaultdict(list)
    for tsv in tsvs:
        df = pd.read_csv(fspath(tsv), sep='\t')
        duplicate_ind = []
        for ind, val in df.duplicated(subset=['participant_id']).items():
            if val: duplicate_ind.append(ind)
        dupes = list(df.loc[df.duplicated(subset=['participant_id']), 'participant_id'])
        if dupes:
            if tsv.name not in combined_forms:
                for onid in dupes:
                    rvid = rvid_lookup(onid.split('-')[1])
                    d[tsv.name].append(f'{onid}: {rvid}')
    with open(filepaths.curr_outputs_dir.joinpath(modality + '_multiple_responses_found.json'), 'w') as f:
        json.dump(d, f, indent=4)


def create_cris_data_dict(df, filename, outdir):
    """create a json data dictionary for a given cris dataframe"""
    json_dict = {}
    for key in df.keys():
        if key != 'participant_id':
            json_dict[key] = {'Units': key.split()[-1]}
        else:
            json_dict[key] = {'Description': 'OpenNeuro ID of the subject.'}
    with open(outdir.joinpath(filename + '.json'), "w") as f:
        json.dump(json_dict, f, indent=4)


def transform_ctdb_legends(dict_of_legends, dict_names_mapping):
    columns = []
    for form_name, legend in dict_of_legends.items():
        df = dict_of_legends[form_name]  # same as `legend`
        columns += list(df.columns)

    header = list(set(columns))
    for form_name, legend in dict_of_legends.items():
        unused_ctdb_legends = {}
        for stdname, values in dict_names_mapping.items():
            if form_name == values['openneuro']:
                print(form_name)

                df = dict_of_legends[form_name]
                df.drop(df.shape[0] - 1)
                out_filename = Path(filepaths.ctdb_out_dir, form_name + '.json')
                qid_uniq = df['QUESTIONID'].unique()  # compile unique qids for the survey
                qids = qid_uniq[qid_uniq != ' '].astype('int')

                d = {}
                d['participant_id'] = {'Description': 'OpenNeuro ID of the subject.'}
                current = ''
                for i, row in df.iterrows():
                    # detecting if on first line of data dictionary/legend
                    if current == '':
                        # start
                        qid = str(int(row['QUESTIONID']))
                        current = qid
                        if pd.isna(row['QUESTION_ALIAS']):
                            ShortName = row['QUESTION_NAME']
                        else:
                            ShortName = row['QUESTION_ALIAS']
                        LongName = row['QUESTION_NAME'] + ' (question ID ' + qid + ')'
                        Description = row['QUESTION_TEXT']
                        Levels = None

                    elif not pd.isna(row['QUESTION_TEXT']):
                        # write
                        d[ShortName] = {}
                        d[ShortName]['LongName'] = LongName
                        d[ShortName]['Description'] = Description
                        if Levels:
                            d[ShortName]['Levels'] = Levels
                        # reset
                        qid = str(int(row['QUESTIONID']))
                        if pd.isna(row['QUESTION_ALIAS']):
                            ShortName = row['QUESTION_NAME']
                        else:
                            ShortName = row['QUESTION_ALIAS']
                        LongName = row['QUESTION_NAME'] + ' (question ID ' + qid + ')'
                        Description = row['QUESTION_TEXT']
                        # DataType = row[10]
                        # Range = row[12]
                        Levels = None

                    if not Levels:
                        if not pd.isna(row['CODEVALUE']):
                            if type(row['DISPLAY']) == float:
                                if pd.isna(row['DISPLAY']):
                                    Levels = {str(int(row['CODEVALUE'])): 'N/A'}
                                else:
                                    Levels = {str(int(row['CODEVALUE'])): str(int(row['DISPLAY']))}
                            else:
                                Levels = {str(int(row['CODEVALUE'])): str(row['DISPLAY'])}
                    else:
                        if not pd.isna(row['CODEVALUE']):
                            if type(row['DISPLAY']) == float:
                                if pd.isna(row['DISPLAY']):
                                    Levels[str(int(row['CODEVALUE']))] = 'N/A'
                                else:
                                    Levels[str(int(row['CODEVALUE']))] = str(int(row['DISPLAY']))
                            else:
                                Levels[str(int(row['CODEVALUE']))] = str(row['DISPLAY'])
                    # detecting if on last line of data dictionary/legend
                    if i == df.shape[0] - 1:
                        # write
                        d[ShortName] = {}
                        d[ShortName]['LongName'] = LongName
                        d[ShortName]['Description'] = Description
                        if Levels:
                            d[ShortName]['Levels'] = Levels
                        with open(fspath(out_filename), 'w') as f:
                            json.dump(d, f, indent=4)
        else:
            unused_ctdb_legends[form_name] = form_name.replace(' ', '_') + '.json'
            with open(fspath(filepaths.curr_outputs_dir.joinpath('unused_ctdb_surveys.json')), 'w') as f:
                f.write(json.dumps(unused_ctdb_legends, indent=4))


def remove_freetext_fields(form_df, excl_fields):
    form_df = form_df.drop(columns=excl_fields, axis=1)
    return form_df


def data_clean_up(dict_of_dfs, outdir):
    """
    - reorder columns so that `participant_id` col comes before all other cols
    - replace NaNs with '-999'
    - replace RV subject ids with openneuro ids
    - remove blank lines
    """
    # rvs_to_be_shared = id_mapping['SUBJECT_NUMBER'].tolist()
    for key, df in dict_of_dfs.items():
        for ind, subjnum in df['participant_id'].items():
            if subjnum in rvs_to_be_shared:
                onid = openneuro_id_lookup(subjnum)
                if onid:
                    dict_of_dfs[key].at[ind, 'participant_id'] = '-'.join(['sub', onid])
            else:
                dict_of_dfs[key] = dict_of_dfs[key].drop(index=ind, axis=0)
        dict_of_dfs[key] = reorder_cols(dict_of_dfs[key])
        dict_of_dfs[key] = remove_blank_rows(dict_of_dfs[key])
    dict_of_dfs = remove_nans(dict_of_dfs)
    return dict_of_dfs


def create_tsvs(modality, dict_of_dfs, outdir):
    """create a tsv file for each dataframe in the dictionary"""
    if modality == 'ctdb':
        # multiple_subj_responses(modality)
        for form_name, df in dict_of_dfs.items():
            df.to_csv(outdir.joinpath(form_name + '.tsv'), sep='\t', quoting=csv.QUOTE_MINIMAL, index=False)

    elif modality == 'cris':
        # multiple_subj_responses(modality)
        for labtest, df in dict_of_dfs.items():
            create_cris_data_dict(dict_of_dfs[labtest], labtest, outdir)
            df.to_csv(fspath(outdir.joinpath(labtest + '.tsv')), sep='\t', quoting=csv.QUOTE_MINIMAL, index=False)


def drop_overlap_subj(revd_df, intersections_file):
    # int_file = Path(filepaths.data_dir, 'overlapping_subjects.txt')
    subjlist = pd.read_csv(fspath(intersections_file), sep=' ', header=None)[0].values.tolist()
    for ind, subj in revd_df['participant_id'].items():
        if subj in subjlist:
            revd_df = revd_df.drop(ind, axis=0)
    return revd_df


def combine_whodas():
    orig_tsv = filepaths.ctdb_out_dir.joinpath('whodas.tsv')
    orig_dict = json.load(open(fspath(filepaths.ctdb_out_dir.joinpath('whodas.json')), 'r'))
    revd_tsv = filepaths.ctdb_out_dir.joinpath('whodas_r.tsv')
    revd_dict = json.load(open(fspath(filepaths.ctdb_out_dir.joinpath('whodas_r.json')), 'r'))

    fields_mapping = {}
    for key, subj in orig_dict.items():
        if key != 'participant_id':
            fields_mapping[subj['LongName'].split()[0]] = key

    # modify revd json dictionary
    old_keys = list(revd_dict.keys())
    for key in old_keys:
        if key != 'participant_id':
            revd_dict[fields_mapping[key]] = revd_dict[key]
            del revd_dict[key]

    # write to json file
    with open(fspath(filepaths.ctdb_out_dir.joinpath('whodas_combined.json')), 'w') as f:
        json.dump(revd_dict, f, indent=4)

    # combine tsvs
    revd_df = pd.read_csv(revd_tsv, sep='\t')
    orig_df = pd.read_csv(orig_tsv, sep='\t')
    old_cols = revd_df.columns
    new_cols = ['participant_id']
    for col in old_cols[1:]:
        new_cols.append(fields_mapping[col])
    revd_df.columns = new_cols

    # drop rows with overlapping subjects
    revd_df = drop_overlap_subj(revd_df,
                                filepaths.data_dir.joinpath('survey_combinations_data/whodas_overlapping_subjs.txt'))
    combined_df = pd.concat([orig_df, revd_df], ignore_index=True)
    combined_df = combined_df.sort_values(by=['participant_id'], ignore_index=True)
    combined_df.to_csv(fspath(filepaths.ctdb_out_dir.joinpath('whodas_combined.tsv')), sep='\t',
                       quoting=csv.QUOTE_MINIMAL, index=False, mode='w')


def combine_demographics():
    orig_df = pd.read_csv(filepaths.ctdb_out_dir.joinpath('demographics.tsv'), sep='\t')
    revd_df = pd.read_csv(filepaths.ctdb_out_dir.joinpath('demographics_r.tsv'), sep='\t')
    # print(revd_df.columns)
    with open(filepaths.data_dir.joinpath('mappings', 'demographics_mappings.json'), 'r') as f:
        mapping = json.load(f)

    orig_df.rename(columns=mapping, inplace=True)
    # print(orig_df.columns)
    orig_df['SETTING'] = orig_df['SETTING'].mask(orig_df['SETTING'] == -999, other=orig_df['Unnamed: 32'])
    del orig_df['Unnamed: 32']
    # print(orig_df)
    revd_df = drop_overlap_subj(revd_df,
                                filepaths.data_dir.joinpath('survey_combinations_data/demo_overlapping_subjs.txt'))

    combined_df = orig_df.append(revd_df, sort=False, ignore_index=True)
    combined_df.fillna(-777, inplace=True)

    for column in ['CURRENT_GENDER_2', 'LGBT_IDENTITY_2', 'RACE_1_2', 'RACE_1_3', 'RACE_1_4',
                   'REFERRAL_TYPE_SECONDARY_4']:
        combined_df[column] = combined_df[column].mask(combined_df[column] == -777, other=-999)
        combined_df[column] = combined_df[column].astype(int)
    combined_df = combined_df.sort_values(by=['participant_id'], ignore_index=True)
    combined_df.to_csv(fspath(filepaths.ctdb_out_dir.joinpath('demographics_combined.tsv')), sep='\t',
                       quoting=csv.QUOTE_MINIMAL, index=False)


def combine_ehi():
    orig_df = pd.read_csv(filepaths.ctdb_out_dir.joinpath('ehi.tsv'), sep='\t')
    revd_df = pd.read_csv(filepaths.ctdb_out_dir.joinpath('ehi_r.tsv'), sep='\t')

    revd_df = drop_overlap_subj(revd_df,
                                filepaths.data_dir.joinpath('survey_combinations_data/ehi_overlapping_subjs.txt'))
    combined_df = pd.concat([orig_df, revd_df], ignore_index=True).fillna(-777)
    for col in combined_df.columns.tolist()[1:]:
        combined_df[col] = combined_df[col].astype(int)
    new_combined_header = \
    pd.read_csv(filepaths.data_dir.joinpath('survey_combinations_data/ehi_columns_order.csv'), header=None)[
        0].values.tolist()
    combined_df.columns = new_combined_header
    combined_df = combined_df.sort_values(by=['participant_id'], ignore_index=True)
    combined_df.to_csv(fspath(filepaths.ctdb_out_dir.joinpath('ehi_combined.tsv')), sep='\t', quoting=csv.QUOTE_MINIMAL,
                       index=False)
